minari/utils/data_collector.py

```diff
@@ -0,0 +1,573 @@
+        self.flatten_observation = check_flatten_space(self.env.observation_space)
+        self.flatten_action = check_flatten_space(self.env.action_space)
+
+    def __call__(
+        self,
+        env: gym.Env,
+        obs: Any,
+        info: Dict,
+        action: Optional[Any] = None,
+        rew: Optional[Any] = None,
+        terminated: Optional[Any] = None,
+        truncated: Optional[Any] = None,
+    ) -> Dict:
+        """Callback method.
+
+        The input arguments belong to a Gymnasium stepping transition: `obs, rew, terminated, truncated, info = env.step(action)`.
+        Override this method to add additional keys or edit each environment's step returns. Additional nested dictionaries can be added to the returned step dictionary
+        as follows:
+
+        ```
+            class CustomStepDataCallback(StepDataCallback):
+                def __call__(self, env, **kwargs):
+                    step_data = super().__call__(env, **kwargs)
+
+                    step_data['environment_states'] = {}
+                    step_data['environment_states']['pose'] = {}
+                    step_data['environment_states']['pose']['position'] = env.position
+                    step_data['environment_states']['pose']['orientation'] = env.orientation
+                    step_data['environment_states']['velocity'] = env.velocity
+
+                    return step_data
+        ```
+
+        The episode groups in the HDF5 file of this Minari dataset will contain a subgroup called `environment_states` with dataset `velocity` and another subgroup called `pose`
+        with datasets `position` and `orientation`
+
+        Args:
+            env (gym.Env): current Gymnasium environment.
+            obs (Any): observation returned by `env.step(action)`
+            info (Dict): information dictionary returned by `env.step(action)`
+            action (Optional[Any], optional): stepping action in `env.step(action)`. Defaults to None.
+            rew (Optional[Any], optional): reward returned by `env.step(action)`. Defaults to None.
+            terminated (Optional[Any], optional): terminated returned by `env.step(action)`. Defaults to None.
+            truncated (Optional[Any], optional): truncated returned by `env.step(action)`. Defaults to None.
+
+        Returns:
+            Dict: dictionary step data. Must contain the keys in STEP_DATA_KEYS = {'actions', 'observations',
+                    'rewards', 'terminations', 'truncations', 'infos'}. Additional key's can be added with nested dictionaries
+        """
+        if action is not None:
+            # Flatten the actions
+            if self.flatten_action:
+                action = spaces.utils.flatten(self.env.action_space, action)
+        # Flatten the observations
+        if self.flatten_observation:
+            obs = spaces.utils.flatten(self.env.observation_space, obs)
+
+        step_data = {
```

Comment:
I would create a `TypedDict` https://docs.python.org/3/library/typing.html#typing.TypedDict for the step data. 
This will improve your type hint as instead of `Dict` you can have `StepData` which is more descriptive to users